### PR TITLE
Migrate nonlocal games 1596 task2 ghz quantum

### PR DESCRIPTION
**Pull request by ggridin**
The link to the issue: https://github.com/microsoft/qsharp/issues/1596

This pull request covering
1. GHZ quantum tasks 2.1 (creating entangled triple).
2. GHZ quantum task 2.2 (quantum strategies).
3. Quantum strategy discussion.
4. GHZ quantum task 2.3 (quantum game) is converted to demo.

### CODE REVIEW

**Comment by ggridin** [comment]
Testing done:

- Build "python ./build.py" is successful
- Placeholder cases are syntactically correct but failing verifications.
- Solutions are successful.
- Some other failure scenarios were tested manually


GHZ Create Entangled Triple
GHZ Create Entangled Triple - Solution
GHZ Quantum Strategy
GHZ Quantum Strategy - Solution
GHZ Discussion 1
GHZ Discussion 2
GHZ Game E2E

**Comment by tcNickolas** [comment]
This week got crazy, so I didn't get to reviewing this - sorry about that! I'll look on Monday, my calendar is a lot clearer then

File: katas/content/nonlocal_games/ghz_quantum_strategy/Solution.qs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            let res = MResetX(qubit);
+            return res == One;
+        }
+        let res = MResetZ(qubit);
+        return res == One;
+    }
+
+    // alternative implementation
+    operation CharlieQuantum (bit : Bool, qubit : Qubit) : Bool {
+        if bit {
```
Comment (tcNickolas) [inline katas/content/nonlocal_games/ghz_quantum_strategy/Solution.qs:22]: Nice way to use the fact that there are three identical code fragments here! (No action required :-))

File: katas/content/nonlocal_games/ghz_quantum_strategy/Verification.qs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    @EntryPoint()
+    operation CheckSolution () : Bool {
+        use qs = Qubit[3];
+        let inputs = RefereeBits();
+        let strategies = [Kata.AliceQuantum, Kata.BobQuantum, Kata.CharlieQuantum];
+
+        let iterations = 1000;
+        mutable wins = 0;
+        for _ in 0 .. iterations - 1 {
```
Comment (tcNickolas) [inline katas/content/nonlocal_games/ghz_quantum_strategy/Verification.qs:42]: Small comment, but since you're not using the value, you can start iterating from 1
```suggestion
        for _ in 1 .. iterations {
```

File: katas/content/nonlocal_games/ghz_quantum_strategy/Verification.qs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                CreateEntangledTriple_Reference(qs);
+                let abc = PlayQuantumGHZ_Reference(strategies, bits, qs);
+                if WinCondition_Reference(bits, abc) {
+                    set wins = wins + 1;
+                }
+		ResetAll(qs);
+            }
+        }
+        if wins < iterations*Length(inputs) {
+            Message($"Player's quantum strategies get {wins} wins out of {iterations*Length(inputs)} possible inputs, which is not optimal");
```
Comment (tcNickolas) [inline katas/content/nonlocal_games/ghz_quantum_strategy/Verification.qs:53]: Technically some of the inputs will be duplicates, so I'd rephrase a bit
```suggestion
            Message($"Players' quantum strategies get {wins} wins out of {iterations*Length(inputs)} runs, which is not optimal");
```

File: katas/content/nonlocal_games/ghz_quantum_strategy/index.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+**Inputs:**
```
Comment (tcNickolas) [inline katas/content/nonlocal_games/ghz_quantum_strategy/index.md:1]: Since this task has three separate operations to implement, I'd suggest phrasing it the way we did in Classical Strategy exercise, clarifying that the learner needs to implement all three and that the inputs to them will be different

File: katas/content/nonlocal_games/ghz_quantum_strategy/solution.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+In Q#, you can perform measurements in a specific basis using either the
+[Measure operation](https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.intrinsic.measure)
```
Comment (tcNickolas) [inline katas/content/nonlocal_games/ghz_quantum_strategy/solution.md:2]: The links to old APIs need to be updated to new

**Review by tcNickolas — CHANGES_REQUESTED** [review]
I looked at the two tasks this PR adds and left some comments. I'll take a look at the discussion and the demo tomorrow, sorry I can't finish it today, so decided to post a partial review. (Though the comment about not defining variable when measuring can be applied to the demo as well)

Thank you!

File: katas/content/nonlocal_games/ghz_create_entangled_triple/index.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+In the quantum version of the game, the players still can not communicate during the game, but they are allowed to share
+qubits from an entangled triple before the start of the game.
+
+**Input:**
+- An array of three qubits in the $\ket{000}$ state.
+
+**Goal:**
+- Create the entangled state $\ket{\Phi} = \frac{1}{2} \big(\ket{000} - \ket{011} - \ket{101} - \ket{110} \big)$ on these qubits.
+This state is equivalent to the [GHZ state]([GHZ state](https://en.wikipedia.org/wiki/Greenberger%E2%80%93Horne%E2%80%93Zeilinger_state))
+$\frac{1}{\sqrt{2}} (\big(\ket{000} + \ket{111} \big)$ up to local unitary operation.
```
[the remark this answers is not in the document]
Comment (ggridin) [reply katas/content/nonlocal_games/ghz_create_entangled_triple/index.md]: Yes, plural (sorry for typo)

File: katas/content/nonlocal_games/ghz_quantum_strategy/index.md (hunk 4 above)
Comment (ggridin) [reply katas/content/nonlocal_games/ghz_quantum_strategy/index.md:4]: Added sentence:
In this task, you should implement three functions, one for each player's quantum strategy.
Note that they are covered by one test, so you must implement all of them to pass the test.

File: katas/content/nonlocal_games/ghz_quantum_strategy/solution.md (hunk 5 above)
Comment (ggridin) [reply katas/content/nonlocal_games/ghz_quantum_strategy/solution.md]: Updated.

File: katas/content/nonlocal_games/ghz_create_entangled_triple/solution.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+1. Apply an X gate to the first and the second qubits to get the $\ket{110}$ state.
+2. Appy an H gate to the first and the second qubits to get the following state:
+$\frac12 \big( \ket{000} - \ket{010} - \ket{100} + \ket{110} \big)$
+3. Flip the sign of the last term using a controlled Z gate with the first qubit as control and the second qubit as target (or vice versa):
+$\frac12 \big( \ket{000} - \ket{010} - \ket{100} -{\color{blue}\ket{110}} \big)$
+4. Now we have the right signs for each term, and the first and the last terms match those of the state we're preparing, so we just need to adjust the two middle terms.
+To do this, we can use [ControlledOnBitString](https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.canon.controlledonbitstring) operation to flip the state of the last qubit if the first two qubits are in $\ket{01}$ or in $\ket{10}$ states, which gives us:
```
[the remark this answers is not in the document]
Comment (ggridin) [reply katas/content/nonlocal_games/ghz_create_entangled_triple/solution.md]: Updated...

File: katas/content/nonlocal_games/ghz_create_entangled_triple/Verification.qs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        Controlled Z([qs[0]], qs[1]);
+
+        // Flip the state of the last qubit for the two middle terms
+        ApplyControlledOnBitString([false, true], X, [qs[0], qs[1]], qs[2]);
+        ApplyControlledOnBitString([true, false], X, [qs[0], qs[1]], qs[2]);
+    }
+
+    @EntryPoint()
+    operation CheckSolution() : Bool {
+        use qs = Qubit[3];
```
[the remark this answers is not in the document]
Comment (ggridin) [reply katas/content/nonlocal_games/ghz_create_entangled_triple/Verification.qs]: Thank you, more clear code now.

File: katas/content/nonlocal_games/ghz_quantum_strategy/Solution.qs [hunk 9]
```diff
@@ -0,0 +1,27 @@
+namespace Kata {
+    operation AliceQuantum (bit : Bool, qubit : Qubit) : Bool {
+        if bit {
+            let res = MResetX(qubit);
+            return res == One;
```
[the remark this answers is not in the document]
Comment (ggridin) [reply katas/content/nonlocal_games/ghz_quantum_strategy/Solution.qs]: All places are updated.

File: katas/content/nonlocal_games/ghz_create_entangled_triple/Solution.qs [hunk 10]
```diff
@@ -0,0 +1,17 @@
+namespace Kata {
+    operation CreateEntangledTriple (qs : Qubit[]) : Unit is Adj {
+        X(qs[0]);
+        X(qs[1]);
+
+        H(qs[0]);
+        H(qs[1]);
+        // At this point we have (|000⟩ - |010⟩ - |100⟩ + |110⟩) / 2
```
[the remark this answers is not in the document]
Comment (ggridin) [reply katas/content/nonlocal_games/ghz_create_entangled_triple/Solution.qs]: Done

File: katas/content/nonlocal_games/ghz_create_entangled_triple/index.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ggridin) [reply katas/content/nonlocal_games/ghz_create_entangled_triple/index.md]: "does it make sense to offer an alternative solution that shows this equivalence?"
This is an interesting question, and **it makes sense to provide proof** of equivalence. 
However, the better place is the following GHZ discussion. I put the reference here,
something like 
"Please refer to the follow-up GHZ Quantum Strategy discussion for details."
and update the discussion a bit later in the next pull request.

**Comment by ggridin** [comment]
> I looked at the two tasks this PR adds and left some comments. I'll take a look at the discussion and the demo tomorrow, sorry I can't finish it today, so decided to post a partial review. (Though the comment about not defining variable when measuring can be applied to the demo as well)
> 
> Thank you!

@tcNickolas - Maria, comments are addressed, waiting for re-review.

File: katas/content/nonlocal_games/index.md [hunk 11]
```diff
@@ -226,6 +226,146 @@ Then, let's proceed with quantum strategy and game implementation.
...
+ 0\\
+-1/2\\
+-1/2\\
+ 0
+\end{bmatrix}$$
+
+Let's first consider the case in which **all three players got the 0 bit**.
+
+When the players make their measurements, they will collectively get one of the basis states of the original state - 000, 011, 101 or 110.
+This measn they'll report back zero \"1\" bits between them (with $25\%$ probability) or two \"1\" bits between them (with $75\%$ probability),
```
Comment (cesarzc) [inline katas/content/nonlocal_games/index.md:279]:
```suggestion
This means they'll report back zero \"1\" bits between them (with $25\%$ probability) or two \"1\" bits between them (with $75\%$ probability),
```

**Review by DmitryVasilevsky — APPROVED** [review]
